chronos/asyncsync.nim

```diff
@@ -638,3 +652,72 @@ proc waitEvents*[T](ab: AsyncEventQueue[T],
         break
 
   events
+
+
+proc newAsyncSemaphore*(size: int = 1): AsyncSemaphore =
+  ## Creates a new asynchronous bounded semaphore ``AsyncSemaphore`` with
+  ## internal available slots set to ``size``.
+  doAssert(size > 0, "AsyncSemaphore initial size must be bigger then 0")
+  AsyncSemaphore(size: size, availableSlots: size)
+
+proc `count`*(s: AsyncSemaphore): int =
+  s.availableSlots
+
+proc tryAcquire*(s: AsyncSemaphore): bool =
+  ## Attempts to acquire a resource, if successful
+  ## returns true, otherwise false
+  ##
+
+  if s.availableSlots > 0 and s.queue.len == 0:
```

Comment:
when is `s.availableSlots > 0` and `s.queue.len != 0`?

Reply:
i think it's the case to acquire without locking (returns immediately), right?